**Introduction: Simulation**

One of the most powerful tools we have in understanding probability is to *simulate* an experiment many times on a computer. We can frequently run thousands or millions of trials and gather a large amount of data regarding the typical outcomes, the average behaviour or properties of a system, and more. In this notebook, you'll start to see how we can use simulation to gather some data. We'll explore a few experiments involving dice rolls.

In this block, we'll roll a set of dice, average the rolls, and see what typical behaviour occurs. One way to "roll a die" numerically is to generate a random number between 0 and 1; then multiply it by 6 and round up to the next integer (we'll actually round down and add 1 because it's a bit easier to code...). Lots of things can be simulated this way, in fact! Generate a random number, choose a particular range of interest, and record a result.

In [16]:
# Import the random number generator package 
from random import random

# Roll one die with a prescribed number of sides
def roll_die(num_sides):
    roll = int(num_sides*random()) + 1
    return roll

# Do this for ten trials and store in a list
rolls = []
num_trials = 100

for _ in range(num_trials):
    rolls.append(roll_die(6))
    
# Print out the results
print(rolls)

[3, 2, 4, 2, 2, 2, 6, 4, 4, 5, 6, 1, 3, 3, 1, 5, 3, 1, 3, 6, 1, 2, 1, 5, 3, 3, 5, 5, 4, 6, 5, 4, 3, 4, 4, 2, 1, 1, 5, 6, 3, 3, 5, 5, 6, 4, 6, 6, 3, 2, 4, 5, 1, 5, 2, 5, 3, 4, 2, 5, 6, 1, 6, 3, 5, 2, 4, 5, 4, 6, 2, 6, 3, 4, 5, 6, 6, 2, 1, 1, 5, 6, 1, 5, 5, 2, 2, 5, 3, 6, 2, 1, 3, 5, 3, 1, 6, 2, 1, 6]


This is just rolling one die ten times, so it's not particularly involved. But suppose we want to return to the first daily homework set, and figure out the probability that two dice add to $8$. We can simulate that here: 

In [2]:
# Do 1000 trials of rolling two dice
num_trials = 1000
success = 0

# Count the number of "successes," i.e. that we get a sum of 8
for _ in range(num_trials):
    r1 = roll_die(6)
    r2 = roll_die(6)

    if r1 + r2 == 8:
        success += 1
        
# Print the estimated probability of success: success / num_trials
print(f'Sucesss: {success}')
print(f'Num Trials: {num_trials}')
print(f'Probability of adding to 8: {success / num_trials}')


Sucesss: 139
Num Trials: 1000
Probability of adding to 8: 0.139


Running $10000$ trials, I had $1338$ successes -- for a probability of $0.1346$. On the other hand, you found that there are $5$ possible dice rolls out of the $36$ total which have a sum of $5$ (2-6, 3-5, 4-4, 5-3, and 6-2). This leads to a probability of $5/36 \approx 0.139$; so not a bad estimate with only ten thousand trials!

Let's make one last modification where we store every outcome and compute the probabilities of each. It's helpful for this part if you know what a [set](https://www.w3schools.com/python/python_sets.asp) is in Python; it's something that pairs a key (here, the value of the sum) with a value (how many times we counted it).

In [3]:
# Do 1000 trials again
num_trials = 10000

# An easy way to store data is to make a set in Python.
success = {r:0 for r in range(2, 13)} 
# This is equivalent to writing {2:0, 3:0, 4:0, ..., 12:0}

for _ in range(num_trials):
    r1 = roll_die(6)
    r2 = roll_die(6)

    s = r1 + r2

    # Update the count of successes
    success[s] += 1

# Now print the results in a bit of a table:
for _ in range(2, 13):
    print(f'Probability of summing to {_} = {success[_] / num_trials}')

Probability of summing to 2 = 0.0277
Probability of summing to 3 = 0.0581
Probability of summing to 4 = 0.0868
Probability of summing to 5 = 0.1076
Probability of summing to 6 = 0.1415
Probability of summing to 7 = 0.1655
Probability of summing to 8 = 0.142
Probability of summing to 9 = 0.1074
Probability of summing to 10 = 0.0774
Probability of summing to 11 = 0.0547
Probability of summing to 12 = 0.0313


**Questions for you**: 

Answer the following questions; in your submission, include any new code that you wrote or modified.

1) Run a few experiments with a small number of trials (on the order of 100). Run a few experiments with a large number of trials (like 100K or more). In which case do you see more variation in the experiment results? Is that what you expected?

In case 1 I see more variation in the expeirment results. In case 2 there is no differnece expect repeating similar results to
case 1. This was not what I expected. I thought more trials would give out more variations, but it seems like all it does
is give me duplicates of case 1. Now that I can see this with my own eyes, my thought process has changed and I will now
expect that the variation will not change much if there is more trials. The true variation is case 1 and case 2 just repeats
it in a longer list.

-(die with a prescribed number of sides): 100 gives a shorter list, 100000 gives a longer list with repeating numbers of 100 list.

-(rolling two dice)Probability of adding: 1000 gives a slightly smaller number 0.135, 200k gives a slightly larger number 0.137735.


2) If you roll three dice instead, what is the most common sum?

If I roll a three dice instead, the most common sum is 9-13. Thus, the true answer I noticed is 10 or 11.
The code is pasted below this section for # 2. It shows the results: This tells me that the most common sums for a
thee dice is 9, 10, 11, 12, and 13. The higher the probalility, the greate change the summ is shown. I noticed that the number that is most likely to gain a sum for a three dice is a 10 or 11. The summ of 9, 12, and 13 flunctuates as it it not always
the same number. For 10 or 11, I notice that the probalility is similar no matter how many times I press run.
Probability of summing to 9 = 0.1128    #flunctuates
Probability of summing to 10 = 0.1283   #about the same
Probability of summing to 11 = 0.124    #about the same
Probability of summing to 12 = 0.1119   #functutates
Probability of summing to 13 = 0.1012   #flunctuates



3) Let's change the experiment: roll two dice and multiply their values instead of adding. Make a table of the probabilities for each outcome in {1, 2, ..., 36}.


See coding below for # 3 Answer




In [7]:
#For question # 2:
# Do 1000 trials again
num_trials = 1000

# An easy way to store data is to make a set in Python.
success = {r:0 for r in range(2, 19)} 
# This is equivalent to writing {2:0, 3:0, 4:0, ..., 12:0}

for _ in range(num_trials):
    r1 = roll_die(6)
    r2 = roll_die(6)
    r3 = roll_die(6)

    s = r1 + r2 + r3

    # Update the count of successes
    success[s] += 1

# Now print the results in a bit of a table:
for _ in range(2, 19):
    print(f'Probability of summing to {_} = {success[_] / num_trials}')

Probability of summing to 2 = 0.0
Probability of summing to 3 = 0.006
Probability of summing to 4 = 0.014
Probability of summing to 5 = 0.028
Probability of summing to 6 = 0.052
Probability of summing to 7 = 0.075
Probability of summing to 8 = 0.097
Probability of summing to 9 = 0.113
Probability of summing to 10 = 0.115
Probability of summing to 11 = 0.129
Probability of summing to 12 = 0.099
Probability of summing to 13 = 0.102
Probability of summing to 14 = 0.075
Probability of summing to 15 = 0.047
Probability of summing to 16 = 0.031
Probability of summing to 17 = 0.013
Probability of summing to 18 = 0.004


In [39]:
#For question #3

# modify similar code for : roll two dice and multiply 
# their values instead of adding. 
# Make a table of the probabilities for each outcome in {1, 2, ..., 36}.
# print out tables and paste code here...
# Do 1000 trials again
num_trials = 1000

# An easy way to store data is to make a set in Python.
success = {r:0 for r in range(1, 37)} 
# This is equivalent to writing {2:0, 3:0, 4:0, ..., 12:0}

for _ in range(num_trials):
    r1 = roll_die(6)
    r2 = roll_die(6)

    s = r1 * r2

    # Update the count of successes
    success[s] += 1

# Now print the results in a bit of a table:
for _ in range(1, 37):
    print(f'Probability of Multiplying to {_} = {success[_] / num_trials}')
    
print(f'\n')


#TABLE SECTION PART 1

# Import the random number generator package 
from random import random

# Roll one die with a prescribed number of sides
def roll_die(num_sides):
    roll = int(num_sides*random()) + 1
    return roll

# Do this for ten trials and store in a list
rolls = []
num_trials = 1000

for _ in range(num_trials):
    rolls.append(roll_die(6)*roll_die(6))
    
# Print out the results
print(f'Table of d*d of outcomes out of 1000 trials:')
print(rolls)

print(f'\n')
print(f'Table of 36 outcomes (no multiplying):')
#TABLE SECTION PART 2

def dice_comb(k, memo={}):
   
    if k == 1:
        memo[1] = [(i,) for i in range(1, 7)]
        return memo[1]
    elif k in memo:
        return memo[k]  
    else:
        prev_res = dice_comb(k-1, memo)
        res = []
        for comb in prev_res:
            for j in range(1, 7):
                res.append(comb + (j,))
        memo[k] = res
        return res
k = 2
print(dice_comb(k))

Probability of Multiplying to 1 = 0.027
Probability of Multiplying to 2 = 0.061
Probability of Multiplying to 3 = 0.057
Probability of Multiplying to 4 = 0.08
Probability of Multiplying to 5 = 0.052
Probability of Multiplying to 6 = 0.108
Probability of Multiplying to 7 = 0.0
Probability of Multiplying to 8 = 0.051
Probability of Multiplying to 9 = 0.031
Probability of Multiplying to 10 = 0.061
Probability of Multiplying to 11 = 0.0
Probability of Multiplying to 12 = 0.109
Probability of Multiplying to 13 = 0.0
Probability of Multiplying to 14 = 0.0
Probability of Multiplying to 15 = 0.055
Probability of Multiplying to 16 = 0.026
Probability of Multiplying to 17 = 0.0
Probability of Multiplying to 18 = 0.061
Probability of Multiplying to 19 = 0.0
Probability of Multiplying to 20 = 0.057
Probability of Multiplying to 21 = 0.0
Probability of Multiplying to 22 = 0.0
Probability of Multiplying to 23 = 0.0
Probability of Multiplying to 24 = 0.05
Probability of Multiplying to 25 = 0.028
Prob